usermods/Tubes/protocol.h

```diff
@@ -0,0 +1,192 @@
+#include "effects.h"
+
+#define CURRENT_NODE_VERSION 2
+
+#pragma pack(push,4) // set packing for consist transport across network
+// ideally this would have been pack 1, so we're actually wasting a
+// number of bytes across the network, but we've already shipped...
+
+#define WLED_MESSAGE_MAX_SIZE 250
+
+typedef enum{
+    RECIPIENTS_ALL=0,  // Send to all neighbors; non-followers will ignore
+    RECIPIENTS_ROOT=1, // Send to root for rebroadcasting downward, all will see
+    RECIPIENTS_INFO=2, // Send to all neighbors "FYI"; none will ignore
+} MessageRecipients;
+
+typedef uint16_t MeshId;
+
+typedef struct {
+    MeshId id = 0;
+    MeshId uplinkId = 0;
+    uint8_t version = CURRENT_NODE_VERSION;
+} MeshNodeHeader;
+
+#define MESSAGE_DATA_SIZE 64
+typedef struct {
+    MeshNodeHeader header;
+    MessageRecipients recipients;
+    uint32_t timebase;
+    CommandId command;
+    uint8_t data[MESSAGE_DATA_SIZE] = {0};
+} NodeMessage;
+
+#pragma pack(pop)
+
+typedef struct wizmote_message {
+    uint8_t program;      // 0x91 for ON button, 0x81 for all others
+    uint8_t seq[4];       // Incremetal sequence number 32 bit unsigned integer LSB first
+    uint8_t byte5 = 32;   // Unknown
+    uint8_t button;       // Identifies which button is being pressed
+    uint8_t byte8 = 1;    // Unknown, but always 0x01
+    uint8_t byte9 = 100;  // Unnkown, but always 0x64
+
+    uint8_t byte10;  // Unknown, maybe checksum
+    uint8_t byte11;  // Unknown, maybe checksum
+    uint8_t byte12;  // Unknown, maybe checksum
+    uint8_t byte13;  // Unknown, maybe checksum
+} wizmote_message;
+
+
+
+#pragma pack(push,1) // set packing for consist transport across network
+
+// This macro is for backwards compatibility with the v2 protocol based on size of messages
+#define WLED_MESSAGE_COMPATIBILITY_ASSERT(msg) static_assert(sizeof(msg) != sizeof(NodeMessage) && sizeof(msg) != sizeof(wizmote_message), # msg " message should not be the same size as NodeMessage")
+
+
+struct WLED_Header {
+
+    enum class ID : uint16_t {
+        None = 0,
+        RTC,
+        GPS,
+        WellknownPeer,
+        LogMessage,
+        Background,
+        Effect,
+        BPM,
+        MAX_ID
+    };
+
+    static_assert(static_cast<uint16_t>(WLED_Header::ID::MAX_ID) <= 0x3FFF, "Too large of message id" );
+
+    uint32_t header_ver = WLEDHeaderValue();
+    uint16_t id:12;
+    uint16_t sig:4;
+    uint8_t  unused = 0;  // perhaps for nonce or checksum  - makes uint32_t aligned
+    uint8_t  len;
+
+    WLED_Header( ID i, uint8_t s = 0) : id(static_cast<uint16_t>(i) & 0x07ff), sig(s & 0x07) {}
+
+    constexpr bool isHeaderValid() {
+        return WLEDHeaderValue() == (header_ver & 0x7f7f7f7f);
+    }
+
+    constexpr uint32_t version() {
+        return header_ver & 0x80808080;
+    }
+
+    constexpr bool isCurrentVersion() {
+        return version() == 0x00000000;
+    }
+
+private:
+    constexpr uint32_t WLEDHeaderValue() {
+        return *((uint32_t*)"WLED");

```

Comment:
all v3 messages start with 4 bytes with the lower 7 bits set to "WLED".  We treat it as a 32 bit number for a faster compare.

The upper bit of each is used for versioning, which gives us room for 15 more future versions